nexus/src/app/crucible.rs

```diff
@@ -91,6 +91,44 @@ impl super::Nexus {
         Ok(!on_in_service_physical_disk)
     }
 
+    /// Return a region's associated address, or None if it can't be retrieved.
+    pub async fn region_addr(
+        &self,
+        log: &Logger,
+        region_id: Uuid,
+    ) -> Result<Option<SocketAddrV6>, Error> {
+        // If a region port was previously recorded, return the address right
+        // away
+
+        if let Some(addr) = self.datastore().region_addr(region_id).await? {
+            return Ok(Some(addr));
+        }
+
+        // Otherwise, ask the appropriate Crucible agent
+
+        let dataset = {
+            let region = self.datastore().get_region(region_id).await?;
+            self.datastore().dataset_get(region.dataset_id()).await?
+        };
+
+        let Some(returned_region) =
+            self.maybe_get_crucible_region(log, &dataset, region_id).await?
+        else {
+            // The Crucible agent didn't think the region exists?
+            // XXX bail! ?
+            warn!(log, "no region for id {region_id}");
+
+            return Ok(None);
+        };
+
+        // Record the returned port
+        self.datastore()
+            .region_set_port(region_id, returned_region.port_number)
+            .await?;
+
+        self.datastore().region_addr(region_id).await
```

Comment:
Should we return an error here if this returns `None`? We *just* called `region_set_port`, so this **should** be returning a known address unless something went wrong, or we concurrently deleted the region.

Then we could presumably change the return type of this whole function to `Result<SocketAddrV6, Error>`, rather than an `Option`.

Reply:
Excellent idea about changing the return type, this is done in ed31cbbda